xonsh/environ.py

```diff
@@ -454,8 +457,9 @@ def multiline_prompt():
     'TITLE': DEFAULT_TITLE,
     'MULTILINE_PROMPT': '.',
     'XONSHRC': os.path.expanduser('~/.xonshrc'),
-    'XONSH_HISTORY_SIZE': 8128,
-    'XONSH_HISTORY_FILE': os.path.expanduser('~/.xonsh_history'),
+    'XONSH_HISTORY_SIZE': (8128, 'files'),
+    'XONSH_HISTORY_FILE': os.path.expanduser('~/.xonsh_history.json'),
+    'XONSH_STORE_STDOUT': True,
```

Comment:
I agree that this is a very nice feature to have, and is potentially something that could set xonsh apart from competitors, but I guess my feeling is still that the best defaults are conservative ones, so I would still advocate not collecting stdout by default.

If we don't want to go that route, though, then I actually think limiting based on number of lines is better (since with a filesize limit, we could theoretically start garbage collecting (and, thus, losing history) after a very small number of commands).

A warning message might be alright, but I think that probably isn't necessary so long as this behavior is well-documented somewhere.  In particular, I think if this "replay" capability is explained in the tutorial (along with `$XONSH_STORE_STDOUT`), then I think we are probably fine.